### PR TITLE
Initial backend implementation for submission interoperability (EXPOSUREAPP-2266)

### PR DESCRIPTION
**Pull request by kolyaopahle**
## Description
This PR implements the new protobuf spec for submission of keys that are federated to the EU Servers.

The base branch is currently equal to a new feature branch, i suggest using this branch to aggregate all the interoperability features.

### CODE REVIEW

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/http/WebRequestBuilder.kt [hunk 1]
```diff
@@ -214,6 +216,8 @@ class WebRequestBuilder(
         val submissionPayload = KeyExportFormat.SubmissionPayload.newBuilder()
             .addAllKeys(keyList)
             .setPadding(ByteString.copyFromUtf8(fakeKeyPadding))
+            .addAllVisitedCountries(visitedCountries)
+            .setConsentToFederation(consentToFederation)
```
Comment (Fabian-K) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/http/WebRequestBuilder.kt:220]: note: the added fields have an impact on plausible deniability. It needs to be checked how this is compensated in the fake submission request

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/SubmissionRepository.kt [hunk 2]
```diff
@@ -67,4 +69,12 @@ object SubmissionRepository {
...
         LocalData.teletan(teletan)
     }
+
+    fun setVisitedCountries(countries: List<String>) {
+        visitedCountries.postValue(countries)
+    }
+
+    fun setConsentToFederation(consent: Boolean) {
+        consentToFederation.postValue(consent)
+    }
```
Comment (Fabian-K) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/SubmissionRepository.kt:79]: Just to double-check: the keyword "persistence" in the commit description does not refer to storing it in LocalData etc, correct?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/viewmodel/SubmissionViewModel.kt [hunk 3]
```diff
@@ -48,11 +49,26 @@ class SubmissionViewModel : ViewModel() {
...
+        return value
+    }
 
     fun submitDiagnosisKeys(keys: List<TemporaryExposureKey>) = viewModelScope.launch {
         try {
             _submissionState.value = ApiRequestState.STARTED
-            SubmissionService.asyncSubmitExposureKeys(keys)
+            SubmissionService.asyncSubmitExposureKeys(
+                notNullGuard(visitedCountries.value),
+                notNullGuard(consentToFederation.value),
```
Comment (Fabian-K) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/viewmodel/SubmissionViewModel.kt:69]: As this feature is not complete I assume that proper handling for the RequiredParameterNotSetException follows in a separate PR

File: Server-Protocol-Buffer/src/main/proto/keyExportFormat.proto [hunk 4]
```diff
@@ -27,9 +27,22 @@ message SignatureInfo {
...
+}
+
 message SubmissionPayload {
-    repeated TemporaryExposureKey keys = 1;
-    optional bytes padding = 2;
+  repeated TemporaryExposureKey keys = 1;
+  optional bytes padding = 2;
+  repeated string visitedCountries = 3;
+  optional string origin = 4;
+  optional ReportType reportType = 5 [default = CONFIRMED_CLINICAL_DIAGNOSIS];
```
Comment (Fabian-K) [inline Server-Protocol-Buffer/src/main/proto/keyExportFormat.proto:44]: origin and reportType will be used in the future?

**Review by Fabian-K — COMMENTED** [review]
Overall this lgtm, just a few comments for clarification :)

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/SubmissionRepository.kt (hunk 2 above)
Comment (kolyaopahle) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/SubmissionRepository.kt:79]: yeah, thats a wording mistake, as of right now the spec contains no definition for additional local storage properties

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/viewmodel/SubmissionViewModel.kt (hunk 3 above)
Comment (kolyaopahle) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/viewmodel/SubmissionViewModel.kt:69]: exacly, this is just placeholder code untill the ui components are built

File: Server-Protocol-Buffer/src/main/proto/keyExportFormat.proto (hunk 4 above)
Comment (kolyaopahle) [reply Server-Protocol-Buffer/src/main/proto/keyExportFormat.proto:44]: origin and report type are part of the current spec but are to be left empty on client side and filled in only on the server backend

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/http/WebRequestBuilder.kt (hunk 1 above)
Comment (kolyaopahle) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/http/WebRequestBuilder.kt:220]: true, will check back with the PO's on how to proceed here

**Review by Fabian-K — DISMISSED** [review]
LGTM given that the impact on plausible deniability is handled separately

**Review by jakobmoellerdev — DISMISSED** [review]
Please resolve Merge conflicts

File: Server-Protocol-Buffer/src/main/proto/applicationConfiguration.proto [hunk 5]
```diff
@@ -13,7 +13,9 @@ message ApplicationConfiguration {
 
     ApplicationVersionConfiguration appVersion = 5;
 
-    repeated string supportedCountries = 6;
+    AppFeatures appFeatures = 6;
+
+    repeated string supportedCountries = 7;
```
Comment (MertSafter) [inline Server-Protocol-Buffer/src/main/proto/applicationConfiguration.proto:18]: Already implemented in https://github.com/corona-warn-app/cwa-app-android/pull/1202

**Comment by harambasicluka** [comment]
Closed as it's implemented in #1202 as mentioned from @MertSafter : https://github.com/corona-warn-app/cwa-app-android/pull/1072#discussion_r492594927

**Pull request closed without merging**